src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -871,48 +877,95 @@ void makeNotNullMembersMaybeNull()
+                    // We want to presume all required members of the type are uninitialized, and in addition we want to set all fields to
+                    // default if we can get to this constructor by doing so (ie, : this() in a value type).
+                    return membersToBeInitialized(method.ContainingType, includeAllMembers: method.IncludeFieldInitializersInBody(), includeRequiredMembers, includeBaseRequiredMembers);
 
-                    static ImmutableArray<Symbol> membersToBeInitialized(NamedTypeSymbol containingType, bool includeAllMembers)
+                    static ImmutableArray<Symbol> membersToBeInitialized(NamedTypeSymbol containingType, bool includeAllMembers, bool includeRequiredMembers, bool includeBaseRequiredMembers)
                     {
-                        var requiredMembers = containingType.AllRequiredMembers;
-                        var members = includeAllMembers ? containingType.GetMembersUnordered() : ImmutableArray<Symbol>.Empty;
-                        if (requiredMembers.IsEmpty)
+                        return (includeAllMembers, includeRequiredMembers, includeBaseRequiredMembers) switch
                         {
-                            return members;
-                        }
+                            (includeAllMembers: false, includeRequiredMembers: false, includeBaseRequiredMembers: false)
+                                => ImmutableArray<Symbol>.Empty,
 
-                        var builder = ArrayBuilder<Symbol>.GetInstance(members.Length + requiredMembers.Count);
-                        builder.AddRange(members);
-                        foreach (var (_, requiredMember) in requiredMembers)
+                            (includeAllMembers: false, includeRequiredMembers: true, includeBaseRequiredMembers: false)
+                                => containingType.GetMembersUnordered().SelectAsArray(predicate: SymbolExtensions.IsRequired, selector: getFieldSymbolToBeInitialized),
+
+                            (includeAllMembers: false, includeRequiredMembers: true, includeBaseRequiredMembers: true)
+                                => containingType.AllRequiredMembers.SelectAsArray(static kvp => getFieldSymbolToBeInitialized(kvp.Value)),
+
+                            (includeAllMembers: true, includeRequiredMembers: _, includeBaseRequiredMembers: false)
+                                => containingType.GetMembersUnordered().SelectAsArray(getFieldSymbolToBeInitialized),
+
+                            (includeAllMembers: true, includeRequiredMembers: true, includeBaseRequiredMembers: true)
+                                => getAllTypeAndRequiredMembers(),
+
+                            (includeAllMembers: _, includeRequiredMembers: false, includeBaseRequiredMembers: true)
+                                => throw ExceptionUtilities.Unreachable,
+                        };
+
+                        ImmutableArray<Symbol> getAllTypeAndRequiredMembers()
```

Comment:
nit: static with `containingType` parameter?

Reply:
I'd prefer to leave these nested as is. They're not useful outside this one method.